File: V5_demo/wss_demo/sbe/level-50/sbe_level_50_intergration.js
const WebSocket = require('ws');
const fs = require('fs');
const crypto = require('crypto');

// -------------------------------------------------------------------
// Config
// -------------------------------------------------------------------

// L50 SBE order book topic
const TOPIC = "ob.50.sbe.BTCUSDT";

// Adjust URL for spot / contract environment as needed:
const WS_URL = "wss://stream-testnet.bybits.org/v5/public-sbe/spot";

// -------------------------------------------------------------------
// Logging setup
// -------------------------------------------------------------------

function setupLogging() {
    const logStream = fs.createWriteStream('logfile_ob50.log', { flags: 'a' });

    function log(level, message) {
        const timestamp = new Date().toISOString();
        const logMessage = `${timestamp} ${level} ${message}\n`;
        logStream.write(logMessage);
        console.log(`[${level}] ${message}`);
    }

    return {
        info: (msg) => log('INFO', msg),
        warning: (msg) => log('WARNING', msg),
        error: (msg) => log('ERROR', msg),
        exception: (error, context) => {
            const msg = `${context}: ${error.message}\n${error.stack}`;
            log('ERROR', msg);
        }
    };
}

const logger = setupLogging();

// -------------------------------------------------------------------
// SBE Parser for OBL50Event (template_id = 20001)
// -------------------------------------------------------------------

class SBEOBL50Parser {
    constructor() {
        // message header: blockLength, templateId, schemaId, version
        this.headerSz = 8; // 4 * uint16 = 8 bytes

        // fixed body fields: 4 x int64 + 2 x int8 + 1 x uint8
        this.bodySz = 4 * 8 + 2 * 1 + 1; // 35 bytes

        // group header for repeating groups: blockLength(uint16), numInGroup(uint16)
        this.groupHdrSz = 4; // 2 * uint16 = 4 bytes

        // each group entry: price(int64), size(int64)
        this.levelSz = 16; // 2 * int64 = 16 bytes

        this.targetTemplateId = 20001;
    }

    // ---------------- core small helpers ----------------

    _parseHeader(data, offset = 0) {
        if (data.length < offset + this.headerSz) {
            throw new Error("insufficient data for SBE header");
        }

        const blockLength = data.readUInt16LE(offset);
        const templateId = data.readUInt16LE(offset + 2);
        const schemaId = data.readUInt16LE(offset + 4);
        const version = data.readUInt16LE(offset + 6);

        return {
            blockLength,
            templateId,
            schemaId,
            version,
            offset: offset + this.headerSz
        };
    }

    _parseVarString8(data, offset) {
        if (offset + 1 > data.length) {
            throw new Error("insufficient data for varString8 length");
        }

        const length = data.readUInt8(offset);
        offset += 1;

        if (length === 0) {
            return { string: "", offset };
        }

        if (offset + length > data.length) {
            throw new Error("insufficient data for varString8 bytes");
        }

        const string = data.toString('utf8', offset, offset + length);
        offset += length;

        return { string, offset };
    }

    _applyExponent(value, exponent) {
        if (exponent >= 0) {
            return Number(value) / Math.pow(10, exponent);
        } else {
            return Number(value) * Math.pow(10, -exponent);
        }
    }

    _parseLevels(data, offset) {
        if (offset + this.groupHdrSz > data.length) {
            throw new Error("insufficient data for group header");
        }

        const blockLength = data.readUInt16LE(offset);
        const numInGroup = data.readUInt16LE(offset + 2);
        offset += this.groupHdrSz;

        if (blockLength < this.levelSz) {
            throw new Error(`blockLength(${blockLength}) < levelSz(${this.levelSz})`);
        }

        const levels = [];
        for (let i = 0; i < numInGroup; i++) {
            if (offset + blockLength > data.length) {
                throw new Error("insufficient data for group entry");
            }

            // Read price and size (first 16 bytes of the block)
            const priceM = data.readBigInt64LE(offset);
            const sizeM = data.readBigInt64LE(offset + 8);
            offset += blockLength; // skip the whole block

            levels.push({
                priceM: priceM,
                sizeM: sizeM
            });
        }

        return { levels, offset };
    }

    // ---------------- public parse ----------------

    parse(data) {
        const headerResult = this._parseHeader(data);
        if (headerResult.templateId !== this.targetTemplateId) {
            //throw new Error(`unsupported template_id=${headerResult.templateId}`);
            return;
        }

        let offset = headerResult.offset;

        if (offset + this.bodySz > data.length) {
            throw new Error("insufficient data for OBL50Event body");
        }

        // Parse fixed body
        const ts = data.readBigInt64LE(offset);
        const seq = data.readBigInt64LE(offset + 8);
        const cts = data.readBigInt64LE(offset + 16);
        const u = data.readBigInt64LE(offset + 24);
        const priceExp = data.readInt8(offset + 32);
        const sizeExp = data.readInt8(offset + 33);
        const pkgType = data.readUInt8(offset + 34);

        offset += this.bodySz;

        // Parse asks group
        const asksResult = this._parseLevels(data, offset);
        const asksRaw = asksResult.levels;
        offset = asksResult.offset;

        // Parse bids group
        const bidsResult = this._parseLevels(data, offset);
        const bidsRaw = bidsResult.levels;
        offset = bidsResult.offset;

        // Parse symbol
        const symbolResult = this._parseVarString8(data, offset);
        const symbol = symbolResult.string;
        offset = symbolResult.offset;

        // Apply exponents
        const asks = asksRaw.map(level => ({
            price: this._applyExponent(level.priceM, priceExp),
            size: this._applyExponent(level.sizeM, sizeExp)
        }));

        const bids = bidsRaw.map(level => ({
            price: this._applyExponent(level.priceM, priceExp),
            size: this._applyExponent(level.sizeM, sizeExp)
        }));

        return {
            header: {
                blockLength: headerResult.blockLength,
                templateId: headerResult.templateId,
                schemaId: headerResult.schemaId,
                version: headerResult.version
            },
            ts: ts.toString(),
            seq: seq.toString(),
            cts: cts.toString(),
            u: u.toString(),
            priceExponent: priceExp,
            sizeExponent: sizeExp,
            pkgType: pkgType, // 0 = SNAPSHOT, 1 = DELTA
            symbol: symbol,
            asks: asks,
            bids: bids,
            parsedLength: offset
        };
    }
}

const parser = new SBEOBL50Parser();

// -------------------------------------------------------------------
// WebSocket handlers
// -------------------------------------------------------------------

function onMessage(data) {
    try {
        if (Buffer.isBuffer(data)) {
            const decoded = parser.parse(data);
            if (decoded !== null && decoded !== undefined) {
                const pkgType = decoded.pkgType;
                const pkgStr = pkgType === 0 ? "SNAPSHOT" : pkgType === 1 ? "DELTA" : `UNKNOWN(${pkgType})`;

                const asks = decoded.asks;
                const bids = decoded.bids;

                const bestAsk = asks.length > 0 ? asks[0] : { price: 0.0, size: 0.0 };
                const bestBid = bids.length > 0 ? bids[0] : { price: 0.0, size: 0.0 };

                const logMessage = `SBE ${decoded.symbol} u=${decoded.u} seq=${decoded.seq} type=${pkgStr} ` +
                    `asks=${asks.length} bids=${bids.length} ` +
                    `BEST bid=${bestBid.price.toFixed(8)}@${bestBid.size.toFixed(8)} ` +
                    `ask=${bestAsk.price.toFixed(8)}@${bestAsk.size.toFixed(8)} ts=${decoded.ts}`;

                logger.info(logMessage);

                console.log(
                    `${decoded.symbol}  u=${decoded.u}  seq=${decoded.seq}  ${pkgStr}  ` +
                    `levels: asks=${asks.length} bids=${bids.length}  ` +
                    `BEST: bid ${bestBid.price.toFixed(8)} x ${bestBid.size.toFixed(8)}  |  ` +
                    `ask ${bestAsk.price.toFixed(8)} x ${bestAsk.size.toFixed(8)}`
                );
            }


        } else {
            // text frame: control / errors / ping-pong
            try {
                const obj = JSON.parse(data);
                logger.info(`TEXT ${JSON.stringify(obj)}`);
                console.log("TEXT:", obj);
            } catch (jsonError) {
                logger.warning(`non-JSON text frame: ${data}`);
                console.log("TEXT(non-json):", data);
            }
        }
    } catch (error) {
        logger.exception(error, "decode error");
        console.log("decode error:", error.message);
    }
}

function onError(error) {
    console.log("WS error:", error);
    logger.error(`WS error: ${error.message}`);
}

function onClose() {
    console.log("### connection closed ###");
    logger.info("connection closed");
}

function pingPeriodically(ws) {
    setInterval(() => {
        try {
            if (ws.readyState === WebSocket.OPEN) {
                ws.send(JSON.stringify({ op: "ping" }));
            }
        } catch (error) {
            console.log("Ping error:", error.message);
        }
    }, 10000);
}

function onOpen(ws) {
    console.log("opened");
    const sub = { op: "subscribe", args: [TOPIC] };
    ws.send(JSON.stringify(sub));
    console.log("subscribed:", TOPIC);

    // Start ping interval
    pingPeriodically(ws);
}

function onPong() {
    console.log("pong received");
}

function onPing() {
    const now = new Date();
    console.log("ping received @", now.toISOString().replace('T', ' ').substring(0, 19));
}

function connWS() {
    const ws = new WebSocket(WS_URL);

    ws.on('open', () => onOpen(ws));
    ws.on('message', onMessage);
    ws.on('error', onError);
    ws.on('close', onClose);
    ws.on('pong', onPing); // Note: In ws library, 'pong' event is for received pongs

    // Set up ping interval
    setInterval(() => {
        if (ws.readyState === WebSocket.OPEN) {
            ws.ping();
        }
    }, 20000);

    return ws;
}

// -------------------------------------------------------------------
// Main execution
// -------------------------------------------------------------------

if (require.main === module) {
    try {
        const ws = connWS();

        // Handle graceful shutdown
        process.on('SIGINT', () => {
            console.log('Shutting down...');
            ws.close();
            process.exit(0);
        });

    } catch (error) {
        console.error('Failed to start WebSocket connection:', error);
        process.exit(1);
    }
}

File: V5_demo/wss_demo/sbe/level-50/sbe_level_50_intergration.py
import json
import logging
import struct
import threading
import time
from datetime import datetime
from typing import Dict, Any, List, Tuple

import websocket

logging.basicConfig(
    filename='logfile_ob50.log',
    level=logging.INFO,
    format='%(asctime)s %(levelname)s %(message)s'
)

# -------------------------------------------------------------------
# Config
# -------------------------------------------------------------------

# L50 SBE order book topic
TOPIC = "ob.50.sbe.BTCUSDT"

# Adjust URL for spot / contract environment as needed:
WS_URL = "wss://stream-testnet.bybits.org/v5/public-sbe/spot"

# -------------------------------------------------------------------
# SBE Parser for OBL50Event (template_id = 20001)
#
# XML schema:
#   ts(int64), seq(int64), cts(int64), u(int64),
#   priceExponent(int8), sizeExponent(int8),
#   pkgType(uint8)   # 0 = SNAPSHOT, 1 = DELTA
#   group asks: blockLen(uint16), numInGroup(uint16),
#               then numInGroup * [ price(int64), size(int64) ]
#   group bids: same as asks
#   symbol(varString8)
# -------------------------------------------------------------------

class SBEOBL50Parser:
    def __init__(self):
        # message header: blockLength, templateId, schemaId, version
        self.header_fmt = "<HHHH"
        self.header_sz = struct.calcsize(self.header_fmt)

        # fixed body fields:
        # ts, seq, cts, u   -> 4 x int64
        # priceExponent, sizeExponent -> 2 x int8
        # pkgType -> uint8
        self.body_fmt = "<qqqqbbB"   # 4*q + 2*b + B
        self.body_sz = struct.calcsize(self.body_fmt)

        # group header for repeating groups: blockLength(uint16), numInGroup(uint16)
        self.group_hdr_fmt = "<HH"
        self.group_hdr_sz = struct.calcsize(self.group_hdr_fmt)

        # each group entry: price(int64), size(int64)
        self.level_fmt = "<qq"
        self.level_sz = struct.calcsize(self.level_fmt)

        self.target_template_id = 20001

    # ---------------- core small helpers ----------------

    def _parse_header(self, data: bytes) -> Dict[str, Any]:
        if len(data) < self.header_sz:
            raise ValueError("insufficient data for SBE header")
        block_length, template_id, schema_id, version = struct.unpack_from(
            self.header_fmt, data, 0
        )
        return {
            "block_length": block_length,
            "template_id": template_id,
            "schema_id": schema_id,
            "version": version,
        }

    @staticmethod
    def _parse_varstring8(data: bytes, offset: int) -> Tuple[str, int]:
        if offset + 1 > len(data):
            raise ValueError("insufficient data for varString8 length")
        (length,) = struct.unpack_from("<B", data, offset)
        offset += 1
        if length == 0:
            return "", offset
        if offset + length > len(data):
            raise ValueError("insufficient data for varString8 bytes")
        s = data[offset: offset + length].decode("utf-8")
        offset += length
        return s, offset

    @staticmethod
    def _apply_exponent(value: int, exponent: int) -> float:
        return value / (10 ** exponent) if exponent >= 0 else value * (10 ** (-exponent))

    def _parse_levels(self, data: bytes, offset: int) -> Tuple[List[Dict[str, float]], int]:
        """
        Parse one repeating group (asks or bids).
        Layout:
           uint16 blockLength
           uint16 numInGroup
           numInGroup * [ price(int64), size(int64) ] (within blockLength)
        """
        if offset + self.group_hdr_sz > len(data):
            raise ValueError("insufficient data for group header")
        block_len, num_in_group = struct.unpack_from(self.group_hdr_fmt, data, offset)
        offset += self.group_hdr_sz

        if block_len < self.level_sz:
            raise ValueError(f"blockLength({block_len}) < level_sz({self.level_sz})")

        levels = []
        for _ in range(num_in_group):
            if offset + block_len > len(data):
                raise ValueError("insufficient data for group entry")
            # we only care about first 16 bytes (price, size)
            price_m, size_m = struct.unpack_from(self.level_fmt, data, offset)
            offset += block_len  # skip the whole block (safe if future adds extra fields)

            levels.append({
                "price_m": price_m,
                "size_m": size_m,
            })
        return levels, offset

    # ---------------- public parse ----------------

    def parse(self, data: bytes) -> Dict[str, Any]:
        hdr = self._parse_header(data)
        if hdr["template_id"] != self.target_template_id:
            raise NotImplementedError(f"unsupported template_id={hdr['template_id']}")

        if len(data) < self.header_sz + self.body_sz:
            raise ValueError("insufficient data for OBL50Event body")

        # parse fixed body
        (ts, seq, cts, u,
         price_exp, size_exp, pkg_type) = struct.unpack_from(
            self.body_fmt, data, self.header_sz
        )

        offset = self.header_sz + self.body_sz

        # asks group
        asks_raw, offset = self._parse_levels(data, offset)
        # bids group
        bids_raw, offset = self._parse_levels(data, offset)
        # symbol
        symbol, offset = self._parse_varstring8(data, offset)

        # apply exponents
        asks = [
            {
                "price": self._apply_exponent(l["price_m"], price_exp),
                "size": self._apply_exponent(l["size_m"], size_exp),
            }
            for l in asks_raw
        ]
        bids = [
            {
                "price": self._apply_exponent(l["price_m"], price_exp),
                "size": self._apply_exponent(l["size_m"], size_exp),
            }
            for l in bids_raw
        ]

        return {
            "header": hdr,
            "ts": ts,
            "seq": seq,
            "cts": cts,
            "u": u,
            "price_exponent": price_exp,
            "size_exponent": size_exp,
            "pkg_type": pkg_type,   # 0 = SNAPSHOT, 1 = DELTA
            "symbol": symbol,
            "asks": asks,
            "bids": bids,
            "parsed_length": offset,
        }


parser = SBEOBL50Parser()

# -------------------------------------------------------------------
# WebSocket handlers
# -------------------------------------------------------------------

def on_message(ws, message):
    try:
        if isinstance(message, (bytes, bytearray)):
            decoded = parser.parse(message)

            pkg_type = decoded["pkg_type"]
            pkg_str = "SNAPSHOT" if pkg_type == 0 else "DELTA" if pkg_type == 1 else f"UNKNOWN({pkg_type})"

            asks = decoded["asks"]
            bids = decoded["bids"]

            best_ask = asks[0] if asks else {"price": 0.0, "size": 0.0}
            best_bid = bids[0] if bids else {"price": 0.0, "size": 0.0}

            logging.info(
                "SBE %s u=%s seq=%s type=%s asks=%d bids=%d "
                "BEST bid=%.8f@%.8f ask=%.8f@%.8f ts=%s",
                decoded["symbol"], decoded["u"], decoded["seq"], pkg_str,
                len(asks), len(bids),
                best_bid["price"], best_bid["size"],
                best_ask["price"], best_ask["size"],
                decoded["ts"],
            )

            print(
                f"{decoded['symbol']}  u={decoded['u']}  seq={decoded['seq']}  {pkg_str}  "
                f"levels: asks={len(asks)} bids={len(bids)}  "
                f"BEST: bid {best_bid['price']:.8f} x {best_bid['size']:.8f}  |  "
                f"ask {best_ask['price']:.8f} x {best_ask['size']:.8f}"
            )

        else:
            # text frame: control / errors / ping-pong
            try:
                obj = json.loads(message)
                logging.info("TEXT %s", obj)
                print("TEXT:", obj)
            except json.JSONDecodeError:
                logging.warning("non-JSON text frame: %r", message)
                print("TEXT(non-json):", message)
    except Exception as e:
        logging.exception("decode error: %s", e)
        print("decode error:", e)


def on_error(ws, error):
    print("WS error:", error)
    logging.error("WS error: %s", error)


def on_close(ws, *_):
    print("### connection closed ###")
    logging.info("connection closed")


def ping_per(ws):
    while True:
        try:
            ws.send(json.dumps({"op": "ping"}))
        except Exception:
            return
        time.sleep(10)


def on_open(ws):
    print("opened")
    sub = {"op": "subscribe", "args": [TOPIC]}
    ws.send(json.dumps(sub))
    print("subscribed:", TOPIC)

    # background ping thread
    threading.Thread(target=ping_per, args=(ws,), daemon=True).start()


def on_pong(ws, *_):
    print("pong received")


def on_ping(ws, *_):
    print("ping received @", datetime.now().strftime("%Y-%m-%d %H:%M:%S"))


def connWS():
    ws = websocket.WebSocketApp(
        WS_URL,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
        on_ping=on_ping,
        on_pong=on_pong,
    )
    ws.run_forever(ping_interval=20, ping_timeout=10)


if __name__ == "__main__":
    websocket.enableTrace(False)
    connWS()